Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,135 @@
+  
+  let asyncBytesIterator: URL.AsyncBytes.AsyncIterator
+  fileprivate let messageFactory: MessageFactory<M>
+  
+  /// The message type in this asynchronous sequence.
+  public typealias Element = M
+  
+  /// An asynchronous iterator that produces the messages of this asynchronous sequence
+  public struct AsyncIterator : AsyncIteratorProtocol, Sendable {
+    
+    public var iter: URL.AsyncBytes.AsyncIterator
+    fileprivate let messageFactory: MessageFactory<M>
+    
+    /// Aysnchronously reads the next varint
+    @inlinable public mutating func nextVarInt() async throws -> UInt64? {
+      var messageSize: UInt64 = 0
+      var shift: UInt64 = 0
+      
+      while let byte = try await iter.next() {
+        messageSize |= UInt64(byte & 0x7f) << shift
+        shift += UInt64(7)
+        if shift > 35 {
+          throw BinaryDecodingError.malformedProtobuf
+        }
+        if (byte & 0x80 == 0) {
+          return messageSize
+        }
+      }
+      if (shift > 0) {
+        // The stream has ended inside a varint.
+        throw BinaryDecodingError.truncated
+      }
+      return nil // End of stream reached.
+    }
+    
+    /// Asynchronously advances to the next message and returns it, or ends the
+    /// sequence if there is no next message.
+    ///
+    /// - Returns: The next message, if it exists, or `nil` to signal the end of
+    ///   the sequence.
+    public mutating func next() async throws -> M? {
+      guard let messageSize = try await nextVarInt() else {
+        return nil
+      }
+      if messageSize == 0 {
+        return try messageFactory.createMesssage(serializedBytes: [])
+      } else if messageSize > 0x7fffffff {
+        throw BinaryDecodingError.tooLarge
+      }
+      
+      var buffer = [UInt8](repeating: 0, count: Int(messageSize))
+      var consumedBytes = 0
+      
+      while let byte = try await iter.next() {
+        buffer[consumedBytes] = byte
+        consumedBytes += 1
+        if consumedBytes == messageSize {
+          return try messageFactory.createMesssage(serializedBytes: buffer)
+        }
+      }
+      throw BinaryDecodingError.truncated // The buffer was not filled.
+    }
+    
+    public typealias Element = M
+  }
+  
+  /// Creates the asynchronous iterator that produces elements of this
+  /// asynchronous sequence.
+  ///
+  /// - Returns: An instance of the `AsyncIterator` type used to produce
+  /// messages in the asynchronous sequence.
+  public func makeAsyncIterator() -> AsyncMessageSequence.AsyncIterator {
+    return AsyncIterator(
+      iter: asyncBytesIterator,
+      messageFactory: messageFactory
+    )
+  }
+}
+
+// Currently neither ExtensionMap nor BinaryDecodingOptions are Sendable
+// hence the @unchecked to suppress warnings
+fileprivate struct MessageFactory<M: Message>: @unchecked Sendable {
+  let extensions: ExtensionMap?
+  let partial: Bool
+  let options: BinaryDecodingOptions
```

Comment:
Yeah we really can't do this. As long as they are not `Sendable` we have should exclude them from the API. I would suggest getting rid of this type altogether for now.

Reply:
Removed the factory in favour of a flattened version (no nested iterator) which avoids the declaration duplication.

Reply:
All params are `Sendable` now.